src/main/cpp/src/datetime_rebase.cu

```diff
@@ -0,0 +1,213 @@
+                                            cudf::detail::copy_bitmask(input, stream, mr),
+                                            input.null_count(), stream, mr);
+
+  thrust::transform(
+      rmm::exec_policy(stream), thrust::make_counting_iterator(0),
+      thrust::make_counting_iterator(input.size()),
+      output->mutable_view().begin<cudf::timestamp_D>(),
+      [d_input = input.begin<cudf::timestamp_D>()] __device__(auto const idx) {
+        auto constexpr julian_end = cuda::std::chrono::year_month_day{
+            cuda::std::chrono::year{1582}, cuda::std::chrono::month{10}, cuda::std::chrono::day{4}};
+        auto constexpr gregorian_start = cuda::std::chrono::year_month_day{
+            cuda::std::chrono::year{1582}, cuda::std::chrono::month{10},
+            cuda::std::chrono::day{15}};
+
+        auto const days_ts = d_input[idx].time_since_epoch().count();
+        auto const days_since_epoch = cuda::std::chrono::sys_days(cudf::duration_D{days_ts});
+
+        // Convert the input into local date in Proleptic Gregorian calendar.
+        auto const ymd = cuda::std::chrono::year_month_day(days_since_epoch);
+        if (ymd > julian_end && ymd < gregorian_start) {
+          // This is the same as rebasing from `ymd = gregorian_start`.
+          return cudf::timestamp_D{cudf::duration_D{-141427}};
+        }
+
+        // No change since this time.
+        if (ymd >= gregorian_start) {
+          return d_input[idx];
+        }
+
+        // Reinterpret year/month/day as in Julian calendar then compute the days since epoch.
+        return cudf::timestamp_D{cudf::duration_D{days_from_julian(ymd)}};
+      });
+
+  return output;
+}
+
+/**
+ * @brief Struct store results of extracting time components from a timestamp.
+ */
+struct time_components {
+  int32_t hour;
+  int32_t minute;
+  int32_t second;
+  int32_t subsecond;
+};
+
+/**
+ * @brief Specialized modulo expression that handles negative values.
+ *
+ * @code{.pseudo}
+ * Examples:
+ *     modulo(1,60)  ->  1
+ *     modulo(-1,60) -> 59
+ * @endcode
+ */
+__device__ __inline__ auto modulo_time(int64_t time, int64_t base) {
+  return static_cast<int32_t>(((time % base) + base) % base);
+}
+
+/**
+ * @brief This function handles converting units by dividing and adjusting for negative values.
+ *
+ * @code{.pseudo}
+ * Examples:
+ *     scale(-61,60) -> -2
+ *     scale(-60,60) -> -1
+ *     scale(-59,60) -> -1
+ *     scale( 59,60) ->  0
+ *     scale( 60,60) ->  1
+ *     scale( 61,60) ->  1
+ * @endcode
+ */
+__device__ __inline__ int64_t scale_time(int64_t time, int64_t base) {
+  return (time - ((time < 0) * (base - 1L))) / base;
+}
+
+int64_t constexpr MICROS_PER_SECOND = 1'000'000L;
+
+__device__ __inline__ time_components get_time_components(int64_t micros) {
+  auto const subsecond = modulo_time(micros, MICROS_PER_SECOND);
+  micros = micros / MICROS_PER_SECOND - ((micros < 0) && (subsecond != 0));
```

Comment:
A comment as to why you're adjusting micros here would be nice. It isn't adjusted for any other `modulo_time` call here.